catalogd/internal/storage/localdir.go

```diff
@@ -20,95 +26,353 @@ import (
+
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	catalogDir := s.catalogDir(catalog)
+	return errors.Join(
+		os.RemoveAll(catalogDir),
+		os.Rename(tmpCatalogDir, catalogDir),
+	)
+}
+
+func (s *LocalDirV1) Delete(catalog string) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	return os.RemoveAll(s.catalogDir(catalog))
+}
+
+func (s *LocalDirV1) ContentExists(catalog string) bool {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalogFileStat, err := os.Stat(catalogFilePath(s.catalogDir(catalog)))
+	if err != nil {
+		return false
+	}
+	if !catalogFileStat.Mode().IsRegular() {
+		// path is not valid content
+		return false
+	}
+
+	if s.EnableQueryHandler {
+		indexFileStat, err := os.Stat(catalogIndexFilePath(s.catalogDir(catalog)))
+		if err != nil {
+			return false
+		}
+		if !indexFileStat.Mode().IsRegular() {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *LocalDirV1) catalogDir(catalog string) string {
+	return filepath.Join(s.RootDir, catalog)
+}
+
+func catalogFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "catalog.jsonl")
+}
+
+func catalogIndexFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "index.json")
+}
+
+type storeMetasFunc func(catalogDir string, metaChan <-chan *declcfg.Meta) error
+
+func storeCatalogData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	f, err := os.Create(catalogFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for m := range metas {
+		if _, err := f.Write(m.Blob); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (s LocalDirV1) Delete(catalog string) error {
-	return os.RemoveAll(filepath.Join(s.RootDir, catalog))
+func storeIndexData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	idx, err := newIndex(metas)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
 }
 
-func (s LocalDirV1) BaseURL(catalog string) string {
+func (s *LocalDirV1) BaseURL(catalog string) string {
 	return s.RootURL.JoinPath(catalog).String()
 }
 
-func (s LocalDirV1) StorageServerHandler() http.Handler {
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
 	mux := http.NewServeMux()
-	fsHandler := http.FileServer(http.FS(&filesOnlyFilesystem{os.DirFS(s.RootDir)}))
-	spHandler := http.StripPrefix(s.RootURL.Path, fsHandler)
-	gzHandler := gzhttp.GzipHandler(spHandler)
 
-	typeHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		w.Header().Add("Content-Type", "application/jsonl")
-		gzHandler.ServeHTTP(w, r)
-	})
-	mux.Handle(s.RootURL.Path, typeHandler)
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
 	return mux
```

Comment:
We should wrap this mix with a middleware hander that checks for HEAD or GET.

Reply:
Something like:

```go
allowedMethodsHandler := func(next http.Handler, allowedMethods ...string) http.Handler {
	allowedMethodSet := sets.New[string](allowedMethods...)
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if !allowedMethodSet.Has(r.Method) {
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}
		next.ServeHTTP(w, r)
	})
}
return allowedMethodsHandler(mux, http.MethodGet, http.MethodHead)
```

Reply:
There are other ways to do it too. For example:
```go
mux.HandleFunc("HEAD "+s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
mux.HandleFunc("GET "+s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
```

And then the same for the query endpoint